packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+     * @param _symbol symbol for the ERC721 token.
+     * @param _baseUri prefix to each ERC721 tokenId's name.
+     * @param _liveness liveness period between submission and verification of a reward.
+     * @param _bondToken ERC20 token that the bond is paid in.
+     * @param _bond size of the bond.
+     * @param _identifier identifier that should be passed to the optimistic oracle on dispute.
+     * @param _customAncillaryData custom ancillary data that should be sent to the optimistic oracle on dispute.
+     * @param _finder finder to look up UMA contract addresses.
+     */
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        string memory _baseUri,
+        uint256 _liveness,
+        IERC20 _bondToken,
+        uint256 _bond,
+        bytes32 _identifier,
+        bytes memory _customAncillaryData,
+        FinderInterface _finder
+    )
+        OptimisticRewarderBase(_liveness, _bondToken, _bond, _identifier, _customAncillaryData, _finder)
+        OptimisticRewarderToken(_name, _symbol, _baseUri)
+    {}
+
+    /**
+     * @notice Used to mint the next ERC721 tokenId.
+     * @param recipient the recipient of the newly minted token.
+     */
+    function mintNextToken(address recipient)
+        public
+        virtual
+        override(OptimisticRewarderBase, OptimisticRewarderToken)
+        returns (uint256)
+    {
+        return OptimisticRewarderToken.mintNextToken(recipient);
+    }
+
+    /**
+     * @notice Used to check the owner of the token.
+     * @dev this override is a formality required by solidity. It forwards the call to the internal ERC721
+     * immplentation.
+     * @param tokenId the tokenId to check the owner of.
+     */
+    function ownerOf(uint256 tokenId) public view virtual override(OptimisticRewarderBase, ERC721) returns (address) {
+        return ERC721.ownerOf(tokenId);
+    }
+}
+
+/**
+ * @notice The optimistic rewarder that does not contain the ERC721 token. It allows the user to pass in an external
+ * ERC721 token.
+ * @dev this setup allows for graceful migrations to new rewarder contracts. It also allows external ERC721 tokens,
+ * like uniswap v3 positions to be rewarded.
+ */
+contract OptimisticRewarderNoToken is OptimisticRewarderBase {
+    OptimisticRewarderToken public token;
+
+    /**
+     * @notice Constructor.
+     * @param _token external ERC721 token for the rewarder to base redemptions on.
+     * @param _liveness liveness period between submission and verification of a reward.
+     * @param _bondToken ERC20 token that the bond is paid in.
+     * @param _bond size of the bond.
+     * @param _identifier identifier that should be passed to the optimistic oracle on dispute.
+     * @param _customAncillaryData custom ancillary data that should be sent to the optimistic oracle on dispute.
+     * @param _finder finder to look up UMA contract addresses.
+     */
+    constructor(
+        OptimisticRewarderToken _token,
```

Comment:
do all NFTs implement the `mintNextToken` interface? how does `mintNextToken` work if this is not implemented? on a uniswap position for example, this is not added, no?

Reply:
There's a subtlety here. If this interface isn't implemented, the call will fail. The effect of this is that all minting functions don't work. However, everything else (including redemptions) will continue to work.

Minting is a non-essential part of the contract, so if you pass in a token that doesn't have that interface/capability, everything should work fine using updateToken to provide any updates you need. Callers, in that setup, should just avoid using that method.

It's admittedly a little hacky, but it is nice to have the option to mint tokens, but allow for use cases where tokens are not mintable or are mintable through custom interfaces.

Reply:
o I see that's pretty awesome. in that case please add additional comments to this effect. this is pretty subtle!

Reply:
Will do!

Reply:
Done.